core/shared/src/main/scala/magnolia.scala

```diff
@@ -326,76 +346,93 @@ object Magnolia {
         val paramsVal = TermName(c.freshName("parameters"))
         val preAssignments = caseParams.map(_.typeclass)
 
-        val defaults = headParamList.fold(List[Tree](q"$scalaPkg.None")) { params =>
-          val x = TermName(c.freshName("x"))
-          val A = TypeName(c.freshName("A"))
-          val dummyArgs = params.filterNot(_.isParamWithDefault).map(p => q"${p.name} = $x")
-          // `A <: Nothing` to avoid dead code warnings
-          val dummy = c.typecheck(q"def $x[$A <: Nothing]($x: $A) = new $genericType(..$dummyArgs)")
-          val defaultParams = dummy.collect { case sel: Select if sel.symbol.isSynthetic => sel }.iterator
-          for (p <- params) yield {
-            if (!p.isParamWithDefault) q"$scalaPkg.None"
-            else q"$scalaPkg.Some(${defaultParams.next()})"
-          }
-        }
-
         val annotations = headParamList.getOrElse(Nil).map(annotationsOf(_))
-        val assignments = caseParams.zip(defaults).zip(annotations).zipWithIndex.map {
-          case (((CaseParam(paramName, repeated, _, paramType, ref), defaultVal), annList), idx) =>
-            val call = if (isValueClass) q"$magnoliaPkg.Magnolia.valueParam" else q"$magnoliaPkg.Magnolia.param"
+        val assignments = if (isReadOnlyTypeclass) {
+          caseParams.zip(annotations).zipWithIndex.map {
+            case ((CaseParam(paramName, repeated, _, paramType, ref), annList), idx) =>
+            val call = if (isValueClass) q"$magnoliaPkg.Magnolia.readOnlyValueParam"
+                       else q"$magnoliaPkg.Magnolia.readOnlyParam"
             q"""$paramsVal($idx) = $call[$typeConstructor, $genericType, $paramType](
+              ${paramName.toString.trim},
+              ${if (isValueClass) q"_.$paramName" else q"$idx"},
+              $repeated,
+              _root_.magnolia.CallByNeed($ref),
+              $scalaPkg.Array($annList)
+            )"""
+          }
+        } else {
+          val defaults = headParamList.fold(List[Tree](q"$scalaPkg.None")) { params =>
+            val x = TermName(c.freshName("x"))
+            val A = TypeName(c.freshName("A"))
+            val dummyArgs = params.filterNot(_.isParamWithDefault).map(p => q"${p.name} = $x")
+            // `A <: Nothing` to avoid dead code warnings
```

Comment:
Really? That works?!

Reply:
Yep 😄 This was just moved though, I did this in #219 and although it still hacky it's less so than what we had before.